### PR TITLE
Refactoring detection modules

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] Organize common functions that are located in each det model into `OTXDetectionModel`
    - `__init__`, `_build_model`, `_create_model`, `_customize_inputs`, `_customize_outputs`, `get_classification_layers`
- [x] Create `common` directory for modules used across tasks
- [x] Remove `DictConfig` to use `dict` instead
- [x] Update docstring
- [x] Refactoring
    - [x] export related things
    - [x] grouped importing backbones, necks, heads, losses
        - `from otx.algo.detection.backbones.cspnext import CSPNeXt` -> `from otx.algo.detection.backbones import CSPNeXt`
    - [x] move assigners, coders, prior_generators, samplers which are located in `heads` to `utils`
        - with this fix, there are only main module directories in `algo/detection`
          <table>
          <tr>
          <td> before </td> <td> after </td>
          </tr>
          <tr>
          <td>

          ```shell
          src/otx/algo/detection \
          ├── backbones
          ├── heads
          │   ├── anchor_generator.py *
          │   ├── anchor_head.py
          │   ├── atss_assigner.py *
          │   ├── atss_head.py
          │   ├── base_head.py
          │   ├── base_sampler.py *
          │   ├── class_incremental_mixin.py
          │   ├── delta_xywh_bbox_coder.py *
          │   ├── distance_point_bbox_coder.py *
          │   ├── dynamic_soft_label_assigner.py *
          │   ├── __init__.py
          │   ├── iou2d_calculator.py *
          │   ├── max_iou_assigner.py *
          │   ├── point_generator.py *
          │   ├── rtmdet_head.py
          │   ├── sim_ota_assigner.py *
          │   ├── ssd_head.py
          │   └── yolox_head.py
          ├── layers
          ├── losses
          ├── necks
          ├── ops
          └── utils
              ├── bbox_overlaps.py
              ├── __init__.py
              ├── structures.py
              └── utils.py
          ```

          </td>
          <td>

          ```shell
          src/otx/algo/detection \
          ├── backbones
          ├── heads
          │   ├── anchor_head.py
          │   ├── atss_head.py
          │   ├── base_head.py
          │   ├── class_incremental_mixin.py
          │   ├── __init__.py
          │   ├── rtmdet_head.py
          │   ├── ssd_head.py
          │   └── yolox_head.py
          ├── layers
          ├── losses
          ├── necks
          └── utils
              ├── assigners
              │   ├── atss_assigner.py *
              │   ├── dynamic_soft_label_assigner.py *
              │   ├── __init__.py
              │   ├── iou2d_calculator.py *
              │   ├── max_iou_assigner.py *
              │   └── sim_ota_assigner.py *
              ├── coders
              │   ├── delta_xywh_bbox_coder.py *
              │   ├── distance_point_bbox_coder.py *
              │   ├── __init__.py
              ├── prior_generators
              │   ├── anchor_generator.py *
              │   ├── __init__.py
              │   ├── point_generator.py *
              ├── samplers
              │   ├── base_sampler.py *
              │   ├── __init__.py
              ├── bbox_overlaps.py
              ├── __init__.py
              ├── nms.py
              ├── structures.py
              └── utils.py
          ```

          </td>
          </tr>
          </table>

### How to test

```python
$ tox -vv -e integration-test-detection
===== 79 passed, 16 skipped, 2 xfailed, 456 warnings in 3297.80s (0:54:57) =====

$ tox -vv -e integration-test-instance_segmentation
===== 47 passed, 24 skipped, 1 xfailed, 267 warnings in 2435.53s (0:40:35) =====
```

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/detection/atss.py [hunk 1]
```diff
@@ -313,15 +155,13 @@ def _build_model(self, num_classes: int) -> SingleStageDetector:
...
-                "nms_pre": 1000,
-            },
-        )
+        test_cfg = {
+            "nms": {"type": "nms", "iou_threshold": 0.6},
+            "min_bbox_size": 0,
+            "score_thr": 0.05,
+            "max_per_img": 100,
+            "nms_pre": 1000,
+        }
```
Comment (harimkang) [inline src/otx/algo/detection/atss.py:162]: My guess is that we'll need to refactor `train_cfg` and `test_cfg` (+other cfg) appropriately in the future as well.

File: src/otx/algo/detection/heads/base_head.py [hunk 2]
```diff
@@ -466,12 +424,14 @@ def export_by_feat(
         bbox_preds: list[Tensor],
         score_factors: list[Tensor] | None = None,
         batch_img_metas: list[dict] | None = None,
-        cfg: DictConfig | None = None,
+        cfg: dict | None = None,
```
Comment (harimkang) [inline src/otx/algo/detection/heads/base_head.py:422]: Is this something that can't be removed? Or how about renaming it something else, because the name `cfg` doesn't really make sense to me.

File: src/otx/algo/detection/utils/samplers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Samplers for detection task."""
+
+from .base_sampler import PseudoSampler, RandomSampler
```
Comment (harimkang) [inline src/otx/algo/detection/utils/samplers/__init__.py]: Do you think these samplers can only be used for detection?

File: src/otx/algo/detection/yolox.py [hunk 4]
```diff
@@ -287,6 +141,10 @@ def _build_model(self, num_classes: int) -> SingleStageDetector:
             num_classes=num_classes,
             in_channels=96,
             feat_channels=96,
+            loss_cls=CrossEntropyLoss(use_sigmoid=True, reduction="sum", loss_weight=1.0),
+            loss_bbox=IoULoss(mode="square", eps=1e-16, reduction="sum", loss_weight=5.0),
+            loss_obj=CrossEntropyLoss(use_sigmoid=True, reduction="sum", loss_weight=1.0),
+            loss_l1=L1Loss(reduction="sum", loss_weight=1.0),
```
Comment (harimkang) [inline src/otx/algo/detection/yolox.py:144]: This change doesn't make sense to me, so I was wondering if you could explain a bit more about why this change is necessary?

File: src/otx/algo/instance_segmentation/mmdet/models/__init__.py
Comment (harimkang) [inline src/otx/algo/instance_segmentation/mmdet/models/__init__.py:1]: In the future, it would be nice to get rid of the folder called mmdet.

**Review by harimkang — DISMISSED** [review]
Thank you for hard work! I have a few quick comments, but just note.

File: src/otx/algo/detection/atss.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/algo/detection/atss.py:162]: I think it should be configurable as well as "image_size", "mean", "std", etc.

File: src/otx/algo/detection/heads/base_head.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/algo/detection/heads/base_head.py:422]: I think we should make it more explicit, not to hold everything in "cfg" as it was in mmdet. Now we can make all required arguments here and eliminate any "cfg"

File: src/otx/algo/detection/utils/assigners/max_iou_assigner.py [hunk 5]
```diff
@@ -1,6 +1,10 @@
 # Copyright (C) 2024 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
-"""Max Iou Assigner implementation from mmdet."""
+# Copyright (c) OpenMMLab. All rights reserved.
+"""Implementations copied from mmdet.models.task_modules.assigners.max_iou_assigner.
```
Comment (kprokofi) [inline src/otx/algo/detection/utils/assigners/max_iou_assigner.py]:
```suggestion
"""Implementation modified from mmdet.models.task_modules.assigners.max_iou_assigner.
```

Comment (kprokofi) [reply src/otx/algo/detection/utils/assigners/max_iou_assigner.py]: I would suggest using "modified" instead of copied everywhere

File: src/otx/algo/instance_segmentation/mmdet/models/__init__.py
Comment (sungchul1) [reply src/otx/algo/instance_segmentation/mmdet/models/__init__.py:1]: Yes, it will be removed through iseg refactoring part.

File: src/otx/algo/detection/yolox.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/algo/detection/yolox.py:144]: I intended to explicitly show what loss functions are used, not entering that class.

File: src/otx/algo/detection/utils/assigners/max_iou_assigner.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algo/detection/utils/assigners/max_iou_assigner.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/3636/commits/216857e0adc42d9df84e0126f8712ae39d269b84

**Review by harimkang — DISMISSED** [review]

File: src/otx/algo/detection/utils/samplers/__init__.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/algo/detection/utils/samplers/__init__.py]: I moved all modules that are used across tasks (det and iseg) to `common` directory (https://github.com/openvinotoolkit/training_extensions/pull/3636/commits/27e15dd35e328719025fbedd09dd75f035765db5).

File: src/otx/algo/detection/heads/base_head.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/base_head.py:422]: I agree with your opinion, but if doing it, the number of arguments will be so many.
Let's talk about this for the next topic.

File: src/otx/algo/detection/atss.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algo/detection/atss.py:162]: I agree with your opinion, but if doing it, the number of arguments will be so many.
Let's talk about this for the next topic.
Before this, I organized arguments that are same with the default value (https://github.com/openvinotoolkit/training_extensions/pull/3636/commits/c04aa58ccb4122de61bdf3561f7c76493114ecd7).

**Review by harimkang — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]
